src/greq/msg.rs

```diff
@@ -0,0 +1,572 @@
+    fn get_aad_slice(&self) -> &[u8] {
+        let self_gva = self as *const _ as *const u8;
+        let algo_gva = &self.algo as *const u8;
+
+        let algo_offset = unsafe { algo_gva.offset_from(self_gva) } as usize;
+
+        unsafe { from_raw_parts(algo_gva, MSG_HDR_SIZE - algo_offset) }
+    }
+
+    /// Get [`SnpGuestRequestMsgHdr`] as a mutable slice reference
+    fn as_slice_mut(&mut self) -> &mut [u8] {
+        unsafe { from_raw_parts_mut(self as *mut _ as *mut u8, MSG_HDR_SIZE) }
+    }
+
+    fn copy_from(&mut self, src: &SnpGuestRequestMsgHdr) {
+        unsafe { copy_nonoverlapping(src, self, 1) };
+    }
+}
+
+impl Default for SnpGuestRequestMsgHdr {
+    // default() method implementation. We can't derive Default because
+    // the field "rsvd3: [u8; 35]" conflicts with the Default trait, which
+    // supports up to [T; 32].
+    fn default() -> Self {
+        Self {
+            authtag: [0; 32],
+            msg_seqno: 0,
+            rsvd1: [0; 8],
+            algo: 0,
+            hdr_version: 0,
+            hdr_sz: 0,
+            msg_type: 0,
+            msg_version: 0,
+            msg_sz: 0,
+            rsvd2: 0,
+            msg_vmpck: 0,
+            rsvd3: [0; 35],
+        }
+    }
+}
+
+#[cfg(not(test))]
+impl Drop for SnpGuestRequestMsg {
+    fn drop(&mut self) {
+        if self.set_encrypted().is_err() {
+            log::error!("SnpGuestRequestMsg: failed to set page to encrypted. Memory leak!");
+            let boxed = unsafe { Box::from_raw(self) };
+            core::mem::forget(boxed);
+        }
+    }
+}
+
+impl SnpGuestRequestMsg {
+    // Allocate the object in the heap without going through stack as
+    // this is a large object
+    //
+    // # Panic
+    //
+    // * Memory allocated is not page aligned or Self does not
+    //   fit into a page
+    pub fn boxed_new() -> Result<Box<Self>, SvsmReqError> {
+        let layout = Layout::new::<Self>();
+
+        // The GHCB spec says it has to fit in one page and be page aligned
+        assert!(layout.size() <= PAGE_SIZE);
+
+        unsafe {
+            let addr = alloc(layout);
+            if addr.is_null() {
+                return Err(SvsmReqError::invalid_request());
+            }
+
+            assert!(VirtAddr::from(addr).is_page_aligned());
+
+            // TODO: have a zeroed page. However, the PSP always returns
+            // the error code 22 (INVALID_PARAM) if the page is written with
+            // zeros here or earlier (e.g. alloc_zeroed, allocate_zeroed_page)
+
+            let ptr = addr.cast::<Self>();
+            Ok(Box::from_raw(ptr))
+        }
+    }
+
+    // Clear the C-bit (memory encryption bit) for the Self page
+    pub fn set_shared(&mut self) -> Result<(), SvsmReqError> {
+        let vaddr = VirtAddr::from(self as *mut Self);
+        this_cpu_mut()
+            .get_pgtable()
+            .set_shared_4k(vaddr)
+            .map_err(|_| SvsmReqError::invalid_request())
```

Comment:
I cannot see where the hypervisor is informed of the encrypted state change for this memory. On my test system this resulted in a lockup when the memory contents were written.

I fixed it with the following temporary code:
```suggestion
        let vaddr = VirtAddr::from(self as *mut Self);
        this_cpu_mut()
            .get_pgtable()
            .set_shared_4k(vaddr)
            .map_err(|_| SvsmReqError::invalid_request())?;

        let paddr = virt_to_phys(vaddr);
        this_cpu_mut()
            .ghcb()
            .page_state_change(
                paddr,
                paddr + PAGE_SIZE,
                PageSize::Regular,
                PageStateChangeOp::PscShared,
            )
            .map_err(|_| SvsmReqError::invalid_request())
```

This also needs to be done in `set_encrypted()` and the corresponding `range` functions in the same module.

Reply:
Arguably, the page state change should be done by the set_shared_4k() and set_encrypted_4k() (set_private_4k ? ) so that every caller of those functions doesn't have to then make the page state change call.

Reply:
For now I am changing the page state like in the code snippet above.

I tried to do that in the `set_shared_4k` and `set_encrypted_4k`, but that causes a `error: kvm run failed Invalid argument`